jit/jit.h

```diff
@@ -0,0 +1,46 @@
+#include <unordered_map>
+#include <atomic>
+#include <mutex>
+
+#include <llvm/IR/LLVMContext.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Target/TargetMachine.h>
+
+namespace msc_jit
+{
+class JITModuleMemoryManager;
+class JITSymbolResolver;
+class JITCompiler;
+class JIT
+{
+ public:
+  JIT();
+  ~JIT();
+  struct CompiledModule
+  {
+    size_t size;
+    uint64_t identifier;
+    std::unordered_map<std::string, void*> function_name_to_symbol;
+  };
+  CompiledModule compileModule(std::function<void(llvm::Module&)> compile_function);
+  void deleteCompiledModule(CompiledModule& module);
+  void registerExternalSymbol(const std::string& symbol_name, void* address);
+
+ private:
+  std::unique_ptr<llvm::Module> createModuleForCompilation();
+  CompiledModule compileModule(std::unique_ptr<llvm::Module> module);
+  static std::unique_ptr<llvm::TargetMachine> getTargetMachine();
+  std::string getMangledName(const std::string& name_to_mangle) const;
+  void runOptimizationPassesOnModule(llvm::Module& module) const;
+  llvm::LLVMContext context;
+  std::unique_ptr<llvm::TargetMachine> target_machine;
+  llvm::DataLayout data_layout;
+  std::unique_ptr<JITCompiler> compiler;
+  std::unique_ptr<JITSymbolResolver> symbol_resolver;
+  std::unordered_map<uint64_t, std::unique_ptr<JITModuleMemoryManager>> module_identifier_to_memory_manager;
+  uint64_t current_module_key = 0;
+  std::atomic<size_t> compiled_code_size;
+  // TODO lock when compiling a module
+  mutable std::mutex jit_lock;
```

Comment:
For any mutex or any potentially concurrently accessed data structure, it would be useful to document the protection rules. Many of these data members are actually protected by `jit_lock` in `deleteCompiledModule()`.